### ISSUE TITLE
SUPPORT: Run control stuck in WAITING

### ISSUE THREAD

**Issue by John-Holt-Tessella** [ISISComputingGroup/IBEX#5771]
LET saw for second time in a day run control stuck in waiting. See #4383 

### Acceptance Criteria

- LET is no longer in waiting
- There is a solution either:
    - bug is found and fixed
    - script is written that runs in background script ioc which recognises when run control has been in waiting for 2 minutes and modifies a block run control settings to be out of range and then reset it. (Maybe a hidden block)
- IS is notified.

**Comment by DominicOram** [comment]
From #4383, when in this state `IN:LET:CS:RC:OUT:LIST` was empty and `IN:LET:CS:RC:INRANGE` was YES but the ICP was still waiting. You can get out of the state by legitimately putting something out of range and then bringing it back in range, this resets the ICP.

**Comment by DominicOram** [comment]
From the archive and genie_python logs order of events are:
* 11:02:22 ICP in SETUP
* 11:02:25 RC:OUT:CNT goes  from 0 to 1
* 11:02:25 RC:OUT:INRANGE goes  from 1 to 0 (Something is out of range)
* 11:02:26 begin sent from genie_python
* 11:02:28 RC:OUT:CNT goes  from 1 to 0
* 11:02:28 RC:OUT:INRANGE goes  from 0 to 1 (Back in range now)
* 11:02:29 ICP in BEGINNING
* 11:02:47 ICP in WAITING

This seems to imply some race condition where going back in range whilst the ICP is beginning is not getting correctly picked up.

**Comment by DominicOram** [comment]
Very similar sequence of events at ~5:24 this morning, was also on the Rot block

**Comment by DominicOram** [comment]
From the logs they're changing the rotation, along with it's runcontrol settings right before the begin, I'm going to suggest they wait for the rotation to finish before doing the begin to avoid the race condition.
```
2020-10-02T11:02:23	(CMD)	(33812)	cset {'kwargs': {'rot': -41.25, 'runcontrol': True, 'lowlimit': -41.35, 'highlimit': -41.15}, 'args': {}}
2020-10-02T11:02:23	(CMD)	(33812)	set_pv_value IN:LET:CS:SB:ROT:SP -41.25
2020-10-02T11:02:23	(CMD)	(33812)	set_pv_value IN:LET:CS:SB:ROT:RC:ENABLE 1
2020-10-02T11:02:23	(CMD)	(33812)	set_pv_value IN:LET:CS:SB:ROT:RC:LOW -41.35
2020-10-02T11:02:23	(CMD)	(33812)	set_pv_value IN:LET:CS:SB:ROT:RC:HIGH -41.15
2020-10-02T11:02:23	(CMD)	(33812)	check_alarms {'kwargs': {}, 'args': ('rot',)}
2020-10-02T11:02:23	(CMD)	(33812)	change_title {'kwargs': {}, 'args': {'title': 'FeTiO3 T_head=1.76K Ei=25meV 240/80 I rot=-41.25 psi=43.5'}}
2020-10-02T11:02:23	(CMD)	(33812)	set_pv_value IN:LET:DAE:TITLE:SP FeTiO3 T_head=1.76K Ei=25meV 240/80 I rot=-41.25 psi=43.5
2020-10-02T11:02:23	(CMD)	(33812)	begin {'kwargs': {}, 'args': {'period': 1, 'meas_id': None, 'meas_type': '', 'meas_subid': '', 'sample_id': '', 'delayed': False, 'quiet': False, 'paused': False, 'verbose': False}}
2020-10-02T11:02:23	(CMD)	(33812)	set_pv_value IN:LET:DAE:PERIOD:SP 1
2020-10-02T11:02:26	(CMD)	(33812)	set_pv_value IN:LET:DAE:BEGINRUNEX 0
```

**Comment by DominicOram** [comment]
Can't easily reproduce this locally. From the logs on an occasion when it worked without a problem we have:

* 12:59:06 ICP in SETUP
* 12:59:07 Block Rot goes out of range
* 12:59:08 begin sent from genie_python
* 12:59:08 ICP in BEGINNING
* 12:59:09 RC:OUT:INRANGE goes  from 1 to 0
* 12:59:09 Block Rot goes in range
* 12:59:11 RC:OUT:INRANGE goes  from 0 to 1
* 12:59:31 ICP in RUNNING

In this case the begin being sent from genie and the ICP beginning has no delay. In the case of something going wrong this begin actually has a ~3 second delay, in which the run control changes. Further investigation is required to see where these ~3 seconds come from but I'm not sure. Any ideas @FreddieAkeroyd?

**Comment by FreddieAkeroyd** [comment]
The ICP looks to have done begin in range, then an "in range" message was received followed by an "out of range" message to leave it waiting. The messages should have been seen in the order "Start SE wait", "Start SE wait".   
```
2020-10-02 11:02:42  *** Timing information for BEGIN: real(12.0444) thread(kernel=0.6875, user=5.6875) process(kernel=3.01563, user=17.1563)
2020-10-02 11:02:43  (0) InstCtrl: [Information] (ISISinstrumentControl::endSEWait) End SE wait
2020-10-02 11:02:44  (0) InstCtrl: [Information] (ISISinstrumentControl::startSEWait) Start SE wait
```
The race condition is that in and out of range use separate PVs (STARTSEWAIT, ENDSEWAIT), both get written to before either can process and then the order they get run in is likely random. Options would be:
* queue requests inside isisdae ioc
* use a single pv and then use fact that CA RPRO feature means you always get latest value sent
There is already a RC:SYNC:SP field in the run control db, it may be triggering this from a scan loop would keep things in sync as a workaround.

**Comment by John-Holt-Tessella** [comment]
I would suggest 2. because 1. would need a wait and how long that wait is will give another race condition.

**Comment by FreddieAkeroyd** [comment]
@John-Holt-Tessella I don't think there is a race condition in 1 any worse than 2, it would queue all incoming request internally until it can process them, so wouldn't order be guaranteed however long it needed to wait?

**Comment by DominicOram** [comment]
Given that we have a new ticket for the actual implementation I'm pointing this for the support call and putting it into review

**Comment by John-Holt-Tessella** [comment]
@FreddieAkeroyd I don't understand what the race conditionis. I thought the start and stop were both sent from the same place (where is this run control ioc?) they are presumably started in order and it is just the rate at which the two PVs get processed and the time between sending and receiving the messages in the isisicp which give rise to the race condition.

**Comment by FreddieAkeroyd** [comment]
@John-Holt-Tessella the run control IOC makes the request to start and stop SE waiting by setting two different PVs in the DAE IOC. I believe the race condition is acquiring the asyn port lock in the DAE IOC, this is acquired by the "begin" and is still held when both the start and stop sample environment wait requests com in. So when the begin completes, the order of waking up "start" and "stop" at the DB side is not preserved. If we queued the start/stop requests we could preserve the original order, but 2. is probably easier to implement. Though actually there is a third option. Run control writes to the start/stop sample environment pv without waiting for them to complete, so it queues both a start and a stop. If it used a completion callback then it would not try to queue a second message until the first had been processed, hence removing the race condition. I probably prefer 2 but completion callbacks might be worth a look.
